### ISSUE TITLE
Remove the option to include / exclude logged in users when the Analytics snippet is not placed via Site Kit

### ISSUE THREAD

**Issue by jamesozzie** [google/site-kit-wp#2910]
## Bug Description

Some Site Kit users may have their Analytics snippet placed manually or via another plugin before connecting the Analytics module within Site Kit. They may have included the logged in user opt-out snippet snippet via another plugin or manually. 

While even the presence of one opt-out snippet will respect the users choice to not include logged in traffic let's consider or start a discussion on removing this option for manually inserted Analytics snippets. This would be useful in particular if there's a possibility that some plugins may insert the opt out snippet below but return a `return false` value. 

```
<!-- Google Analytics user opt-out added via Site Kit by Google -->
 <script type="text/javascript">window["_gaUserPrefs"] = { ioo : function() { return true; } }</script>
```

When that when connecting the Analytics module initially there are no additional configuration options visible - so a user would not be aware of the opt-out snippet (See #2909)

Some considerations for any changes to this configuration:
- Should this configuration option be removed for when users insert Analytics via Tag Manager
- Should these options be visible when initially setting up the Analytics module if a snippet was manually placed

## Steps to reproduce

1. Insert a manually placed Site Kit snippet
2. Connect the Analytics module within Site Kit
3. No configuration options appear but the opt out snippet exists - placed via SK
4. A user must go into their SK Analytics settings to toggle off this snippet

## Screenshots

![ex](https://user-images.githubusercontent.com/41326532/109727132-b880b980-7bab-11eb-828b-46b186edc30e.gif)

![image](https://user-images.githubusercontent.com/41326532/109715158-36889480-7b9b-11eb-9f65-f5e6c35c0a99.png)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* If Site Kit does not place the Analytics snippet (neither via Analytics nor via Tag Manager), the option to exclude certain users from tracking should not be displayed.
* Similarly, in that case the respective snippet should not be rendered in the frontend.
    * While it is technically possible to render it even when the snippet comes from elsewhere, this is more confusing than helpful, plus it might conflict with e.g. the other plugin already having that feature on its own.

## Implementation Brief

* Edit `assets/js/modules/analytics/components/settings/SettingsForm.js`:
  * Grab the `useSnippet` setting of Analytics and Tag Manager modules using `getUseSnippet` selectors.
  * Grab Analytics property ID from the Tag Manager using the `getSingleAnalyticsPropertyID` selector from the `modules/tag-manager` datastore.
  * Conditionally display the `<TrackingExclusionSwitches />` component if:
    * The `useSnippet` of the Analytics module is `true`
    * Or the property ID from the tag manager is not empty **and** Tag Manager's `useSnippet` is `true`.
* Edit the `Analytics::is_tracking_disabled` method:
  * At the very beginning of the function, apply a new filter `googlesitekit_allow_tracking_disabled` with the default value of the `useSnippet` setting of the Analytics module.
  * If this filter returns falsy result, then return `false` immediately and finish the function.
* Edit the `Tag_Manager` class:
  * Add a new private method `filter_analytics_allow_tracking_disabled( $allowed )`.
    * If `$allowed` already equals `true`, do nothing and return early.
    * Return `true` if Tag Manager's `useSnippet` setting is `true` and `gaPropertyID` is not empty.
    * Otherwise return the original value of the `$allowed` argument.
  * Add a new hook for the `googlesitekit_allow_tracking_disabled` filter in the `register` method that proxies request to the `filter_analytics_allow_tracking_disabled` method.

### Test Coverage

* Update existing tests if any is failing due to the aforementioned changes.

### Visual Regression Changes

* N/A

## QA Brief

* Follow the steps to reproduce above
* Verify that the opt out snippet does not appear
* Verify that the option to exclude users from tracking is not displayed in Analytics settings.

## Changelog entry

* Remove the option to include/exclude logged-in users when the Analytics snippet is not placed via Site Kit.

**Comment by eugene-manuilov** [comment]
@Hazlitte 

> * edit assets/js/modules/analytics/components/settings/SettingsForm.js
> * add a new useSnippet variable with a value of useSelect( ( select ) => select( STORE_NAME ).getUseSnippet() );
> * Conditionally display the <TrackingExclusionSwitches /> component if useSnippet is true

We also need to take into account the current Tag Manager setup because it may contain a container that has Google Analytics configuration as well (that's why AC mentions Tag Manager). Check the `useExistingTagEffect` hook to see how we get `propertyID` from Tag_Manager module: https://github.com/google/site-kit-wp/blob/299cb89e1f0c807d6666fd16c9130cd6dbb80536/assets/js/modules/analytics/hooks/useExistingTagEffect.js#L64-L65
In other words, we need to display `<TrackingExclusionSwitches />` element either if `useSnippet` is true or the Tag Manager has non empty `propertyID`. @felixarntz am I right? :thinking: 

> * edit includes/Modules/Analytics.php
> * add an is_useSnippet() function similar to is_tracking_disabled which returns the value for that option
> * modify the print_tracking_opt_out() function to return and not output the snippet if useSnippet is false

Instead of introducing a new method, I think we should modify `is_tracking_disabled` to take into account the current `useSnippet` setting value. Plus, we need to check Tag Manager settings as well leveraging `googlesitekit_analytics_tracking_disabled` filter.

**Comment by Hazlitte** [comment]
@eugene-manuilov Thank you for your comment.  I have updated in the IB accordingly.

**Comment by eugene-manuilov** [comment]
Thanks, @Hazlitte. This issue is a bit tricky, so I have updated your IB to provide more details.

**Comment by Hazlitte** [comment]
👍

**Comment by aaemnnosttv** [comment]
@eugene-manuilov given some of this logic is already in place for Analytics via `canUseSnippet`, I think we can simplify it a little bit by referencing that setting in the logic. Then maybe we don't even need to hook into it (again) from GTM?

**Comment by eugene-manuilov** [comment]
@aaemnnosttv as far as I see it, we can't rely on the logic for `canUseSnippet` because it has different conditions. For example, `canUseSnippet` will be `false` if we have a GTM container with an Analytics tag. If we relied on `canUseSnippet`, then we would have to hide the tracking switch which is wrong.

**Comment by aaemnnosttv** [comment]
> If we relied on `canUseSnippet`, then we would have to hide the tracking switch which is wrong.

@eugene-manuilov  The problem with `canUseSnippet` and relying on GTM setting for `propertyID` at all is that we don't know if the container is rendered by Site Kit or an existing tag by the settings alone. Even though GTM has a `useSnippet` setting of its own, we don't currently disable it if there is an existing tag (it's only controlled by `UseSnippetSwitch`). Looks like this is a bug because it results in a double tag on the front end.

@felixarntz – do you want to fix this as part of this issue or handle that separately in a new issue?

> Add a new private method `filter_analytics_tracking_disable( $disabled, $analytics_settings )`.
>     * It should return the original value of the `$disabled` if it equals `true`.
>     * Return `false` if:

This is a bit confusing – if it returns `true` if `$disabled` is `true`, then the "else" there would always return `false` regardless of the other conditions so it seems like this filter wouldn't change the value.

I think the intention of this filter should be to set the value from `true` (disabled) to `false` (normal non-opt-out behavior) if GTM is set up with an Analytics property and the snippet is enabled, otherwise return the given value. The IB currently ignores this by preserving the `true` value if passed.

**Comment by felixarntz** [comment]
@aaemnnosttv Let's fix this problem via a separate issue.

Regarding conditional visibility of the field, I think the existing IB makes sense here; if there is a single GTM GA property ID, it means GTM is used for the Analytics snippet.

On the server-side, how about we simplify the approach via a new filter instead of modifying the existing one? The purpose of the new functionality here is slightly different, and adding the new `$analytics_settings` to the existing filter just to make this work seems a bit too clever to me. I think a solution that's much easier to understand would be something like:
* `Analytics::is_tracking_disabled` first applies a new filter `googlesitekit_allow_tracking_disabled` (or another name along these lines), with the default value being the `useSnippet` setting value.
* If the filter value is `false`, the entire function bails early, before even looking at the `trackingDisabled` setting.
* `Tag_Manager` then uses that filter: If the filter value isn't already `true`, it checks whether there's a `gaPropertyID` in GTM and if so, sets it to `true`.

**Comment by eugene-manuilov** [comment]
Thanks, @aaemnnosttv and @felixarntz. I have updated the IB. Could you please check it on your end and let me know if it looks good now?

**Comment by eugene-manuilov** [comment]
Thanks, @aaemnnosttv. IB is updated.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

- Verified that the opt out snippet does not appear
- Verified that the option to exclude users from tracking is not displayed in Analytics settings.

![image](https://user-images.githubusercontent.com/73545194/123125115-25aca900-d440-11eb-9ffd-eb65aafb8b61.png)
